python/ray/_raylet.pyx

```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
     return TaskID(object_id.native().TaskId().Binary())
 
 
+def setup_logging(stdout_name, stderr_name):
+    """Sets up logging for the current worker, creating the file and flushing
+    buffers as is necessary.
+
+    Args:
+        stdout_name (str): The file name that stdout should be written to.
+        stderr_name(str): The file name that stderr should be written to.
+
+    Returns:
+        (tuple) The absolute paths of the files that stdout and stderr will be
+    written to.
+    """
+    stdout_path = ""
+    stderr_path = ""
+    worker_pid = os.getpid()
+
+    cdef extern from "ray/util/util.h":
+        void flush_out()
+        void flush_err()
+
+    if stdout_name:
+        # Line-buffer the output (mode 1).
+        stdout_file = open_worker_log(stdout_name, worker_pid)
+
+        # Before redirecting stdout we need to make sure userspace buffers
+        # are all flushed.
+        # Flush python's userspace buffers
+        sys.stdout.flush()
+
+        # TODO (Alex): Flush the c/c++ userspasce buffers if necessary.
+        # `fflush(stdout); cout.flush();`
+
+        stdout_fileno = sys.stdout.fileno()
+
+        # Redirect stdout at the file descriptor level. If we simply set
+        # sys.stdout, then logging from C++ can fail to be
+        # redirected. Note that dup2 will automatically close the old file
+        # descriptor before overriding it.
+        os.dup2(stdout_file.fileno(), stdout_fileno)
+
+        # We must close the original file descriptor to avoid leaking file
+        # descriptors.
+        stdout_file.close()
+
+        # We also manually set sys.stdout and sys.stderr because that seems to
+        # have an affect on the output buffering. Without doing this, stdout
+        # and stderr are heavily buffered resulting in seemingly lost logging
+        # statements. We never want to the stdout file descriptor, dup2 will
+        # close it when necessary and we don't want python's GC to close it.
```

Comment:
There's a lot going on here and I think some of it is confusing. Setting a field like `sys.stdout` doesn't inherently affect buffering (buffering is a property of the stream itself). Furthermore, the default stdio buffering behavior is affected by whether Python is started via `python -u`, so that's one way to disable buffering for spawned processes (such as workers). If you're trying to do this for the driver, though... I feel like it's the caller's responsibility to manage stdio, and to ensure they're redirected before the process is spawned. It seems a bit unexpected to pull the rug out from underneath the caller, since the caller (which may be the user) might deliberately set up stdio in some particular way. It might be good to discuss the approach to see if there might be a better way?

Reply:
I haven't looked super closely into why this is necessary. This is refactored in from [here](https://github.com/ray-project/ray/blob/6c49c01837ef9d83b998a2ec1a269a1a6767fc66/python/ray/worker.py#L1172) (perhaps @robertnishihara remembers). 

Existing behavior is we want stdout to have `open(..., buffering=1)` so we probably want to do that at least once.

Reply:
Ohh I see. We can probably ignore changing the approach for this PR then, it didn't hit me it was what the code was already doing in `worker.py`.